sound/soc/sof/intel/hda-dsp.c

```diff
@@ -338,49 +338,154 @@ static int hda_dsp_send_pm_gate_ipc(struct snd_sof_dev *sdev, u32 flags)
 				  sizeof(pm_gate), &reply, sizeof(reply));
 }
 
-int hda_dsp_set_power_state(struct snd_sof_dev *sdev,
-			    enum sof_d0_substate d0_substate)
+int hda_dsp_set_d0i0(struct snd_sof_dev *sdev)
 {
+	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
 	struct hdac_bus *bus = sof_to_bus(sdev);
 	u32 flags;
-	int ret;
-	u8 value;
+	int ret = 0;
+
+	mutex_lock(&hda->ps_mutex);
+	/* return if not in D0I3 */
+	if (hda->dsp_power_state != SOF_DSP_STATE_D0I3) {
+		dev_vdbg(bus->dev, "%s dsp power state already in %d\n",
+			 __func__, hda->dsp_power_state);
+		goto exit;
+	}
 
 	/* Write to D0I3C after Command-In-Progress bit is cleared */
 	ret = hda_dsp_wait_d0i3c_done(sdev);
 	if (ret < 0) {
 		dev_err(bus->dev, "CIP timeout before D0I3C update!\n");
-		return ret;
+		goto exit;
 	}
 
 	/* Update D0I3C register */
-	value = d0_substate == SOF_DSP_D0I3 ? SOF_HDA_VS_D0I3C_I3 : 0;
-	snd_hdac_chip_updateb(bus, VS_D0I3C, SOF_HDA_VS_D0I3C_I3, value);
+	snd_hdac_chip_updateb(bus, VS_D0I3C, SOF_HDA_VS_D0I3C_I3, 0);
 
 	/* Wait for cmd in progress to be cleared before exiting the function */
 	ret = hda_dsp_wait_d0i3c_done(sdev);
 	if (ret < 0) {
 		dev_err(bus->dev, "CIP timeout after D0I3C update!\n");
-		return ret;
+		goto exit;
 	}
 
 	dev_vdbg(bus->dev, "D0I3C updated, register = 0x%x\n",
 		 snd_hdac_chip_readb(bus, VS_D0I3C));
 
-	if (d0_substate == SOF_DSP_D0I0)
-		flags = HDA_PM_PPG;/* prevent power gating in D0 */
-	else
-		flags = HDA_PM_NO_DMA_TRACE;/* disable DMA trace in D0I3*/
+	flags = HDA_PM_PPG;/* prevent power gating in D0 */
 
 	/* sending pm_gate IPC */
 	ret = hda_dsp_send_pm_gate_ipc(sdev, flags);
-	if (ret < 0)
+	if (ret < 0) {
 		dev_err(sdev->dev,
 			"error: PM_GATE ipc error %d\n", ret);
+		goto exit;
+	}
+
+	/* success, update power state */
+	hda->dsp_power_state = SOF_DSP_STATE_D0;
 
+exit:
+	mutex_unlock(&hda->ps_mutex);
 	return ret;
 }
 
+int hda_dsp_set_d0i3(struct snd_sof_dev *sdev, u32 flags)
```

Comment:
There's too  much in common between set D0i0 and set_D0i3. I think we can unify these 2 functions.